src/common/ult_stack_mmap.c

```diff
@@ -0,0 +1,612 @@
+/**
+ * (C) Copyright 2021-2024 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * \file
+ *
+ * This file implements an alternate/external way for ULTs stacks allocation.
+ * It is based on mmap() of MAP_STACK|MAP_GROWSDOWN regions, in order to
+ * allow overrun detection along with automatic growth capability.
+ */
+
+#define D_LOGFAC DD_FAC(stack)
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include <errno.h>
+#include <string.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <search.h>
+#include <sys/mman.h>
+
+#include <daos/common.h>
+#include <daos_srv/daos_engine.h>
+#include <daos/daos_abt.h>
+#include <daos/ult_stack_mmap.h>
+
+#define MMAP_ULT_STACK_PROT     PROT_READ | PROT_WRITE
+#define MMAP_ULT_STACK_FLAGS    MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE
+#define MMAP_ULT_STACK_FD       -1
+#define MMAP_ULT_STACK_OFFSET   0
+
+#define MREMAP_ULT_STACK_PROT   PROT_READ | PROT_WRITE
+#define MREMAP_ULT_STACK_FLAGS  MAP_PRIVATE | MAP_ANONYMOUS | MAP_STACK | MAP_FIXED | MAP_NORESERVE
+#define MREMAP_ULT_STACK_FD     -1
+#define MREMAP_ULT_STACK_OFFSET 0
+
+#define MMAP_GUARD_PAGE_PROT    PROT_NONE
+#define MMAP_GUARD_PAGE_FLAGS   MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED | MAP_NORESERVE
+#define MMAP_GUARD_PAGE_FD      -1
+#define MMAP_GUARD_PAGE_OFFSET  0
+
+/** Methods of ULT thread creation */
+enum thread_create_flags {
+	/** Create a new ULT and schedule it in a given pool */
+	TCF_ON_POOL = (1 << 0),
+	/** Create a new ULT associated with an execution stream */
+	TCF_ON_XSTREAM = (1 << 1)
+};
+
+/** Pool of mmap ULT free stack */
+struct stack_mmap_pool {
+	/** Binary tree holding list of ULT stacks */
+	void      *smp_rec_btree;
+	/** Lock protecting the binary tree smp_rec_btree */
+	ABT_rwlock smp_rec_btree_rwlock;
+	/** ABT_key used for calling the callback free_mmap_cb() */
+	ABT_key    smp_free_mmap_key;
+	/** ABT thread default stack size */
+	size_t     smp_thread_stack_size;
+	/** Size of memory page */
+	size_t     smp_page_size;
+};
+
+/** Record of a btree holding a list of mmap ULT free stack with a given size */
+struct stack_mmap_rec {
+	/** List of free stacks */
+	d_list_t  smr_desc_list;
+	/** Mutex protecting the list of available stack */
+	ABT_mutex smr_desc_list_mutex;
```

Comment:
Same question here.

Reply:
Same issue here as the list of availabe stack is not per xstream.